Bluetooth: host: Split HCI core to separate modules

joerchan opened the pull request:
Split hci_core core to separate modules:

- br: Mainly discovery and inquiry handling (SSP has its own module)
- adv: Legacy, Extended and periodic advertiser
- scan: Scan and periodic advertising sync.
- id: Identity management, identity address (multiplexing of the random addr) privacy handling (hiding of identity) and identity resolving key handling in the controller.

joerchan committed:
Bluetooth: host: Refactor out identity handling from hci_core to id

Refactor out the identity handling from hci_core to its own source file
in id.c
Identity consistes of managing the identities of the device, the privacy
feature which hides the identities. And handling of the identity
resolving list in the controller, needed to support privacy-enabled
remote devices.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Refactor out scan roles from hci_core to scan

Refactor out the scan roles handling from hci_core.c to its own source
file in scan.c.
Scan roles consists of regular scanning and synchronization to
periodic advertiser.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Refactor out advertiser roles from hci_core to adv

Refactor out the advertiser roles handling from hci_core.c to its own
source file in adv.c.
Advertising roles consists of legacy and extended advertiser, and
periodic advertiser.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Refactor command state set handling to one function

Refactor command state state handling to one function.
When setting state outside of hci_core the cmd macro is not available,
so in order to assign the buf pointer to state the function is needed.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix compilation error with BR/EDR enabled

Fix compilation error with BR/EDR is enabled but the feature
CONFIG_BT_KEYS_OVERWRITE_OLDEST is disabled.
This caused compilation error because the field aging_counter is not
defined in the keys struct.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Align SSP HCI event handler function names

Align the SSP HCI event handlers with the rest of the functions in using
the bt_ prefix. In order to avoid name conflict with the HCI struct
definitions the infix _evt_ has been dropped.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Refactor out BR/EDR from hci_core to br

Refactor out the BR/EDR handling from hci_core.c to its own source file
in br.c.
BR/EDR consists of inquiry and discovery roles in addition to
initialization. SSP is still kept separate.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix style warnings in existing code

Fix style warnings in existing code.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/id.h:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2015-2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Similar comment regarding copyright for a new file

Thalley reviewed (changes requested):
Only reviewed the first two commits. 

To be honest, it's quite hard to figure out what code was moved, and what code is new, and how much attention should be brought to each function.

joerchan commented on subsys/bluetooth/host/id.c:
```diff
@@ -0,0 +1,1740 @@
+}
+
+#if defined(CONFIG_BT_PRIVACY)
+static void adv_is_private_enabled(struct bt_le_ext_adv *adv, void *data)
```
It's just a complicated way to return, since the function has to follow the adv_foreach declaration, then the state contains the value returned.

This code is just moved.

joerchan committed:
Bluetooth: host: Dont include bt_adv_lookup_legacy when adv is disabled

Refactor the code so that bt_adv_lookup_legacy is always called after
the IS_ENABLED(CONFIG_BT_BROADCASTER) configuration has been checked and
the code path will be excluded.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -137,9 +137,13 @@ static struct cmd_data cmd_data[CONFIG_BT_HCI_CMD_COUNT];
 
-void bt_hci_cmd_data_state_set(struct net_buf *buf,
-			       struct bt_hci_cmd_state_set *state)
+void bt_hci_cmd_state_set_init(struct net_buf *buf,
```
I thought keeping the _init variant made most sense, since it is all just initializing the state change when the command completes, now it just includes assigning the pointer from the buf to the state change.

Also the _data_ part in `bt_hci_cmd_data_state_set` was there just to avoid a name conflict with the struct itself, which is another reason why I preferred the _init name.

joerchan commented:
Pushed a fix to the compilation error for the undefined function bt_adv_lookup_legacy.
made it so that it is a refactor commit before the code was moved, so that it remains a clean move.
Changes are in the "Bluetooth: host: Dont include bt_adv_lookup_legacy when adv is disabled" commit.

joerchan committed:
Bluetooth: host: Don't include adv checks when adv is disabled

Don't include checks for advertiser is enabled when the broadcaster role
is not enabled. This makes it possible to have bt_le_ext_adv_foreach
be excluded from the compilation.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
Pushed a fix to the compilation error for undefined definition of bt_le_ext_adv_foreach, now that it is excluded from the build.
This is made as a single commit after id.c had been created but before adv.c was created.
Commit: Bluetooth: host: Don't include adv checks when adv is disabled

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -562,2355 +511,1756 @@ static int set_le_adv_enable_ext(struct bt_le_ext_adv *adv,
 
-static int set_adv_random_address(struct bt_le_ext_adv *adv,
-				  const bt_addr_t *addr)
+static void adv_id_check_connectable_func(struct bt_le_ext_adv *adv, void *data)
```
Only renamed functions that are non-static and moved to header files. So will leave these alone.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -562,2355 +511,1756 @@ static int set_le_adv_enable_ext(struct bt_le_ext_adv *adv,
-		return 0;
+	if (enable == BT_HCI_LE_SCAN_ENABLE) {
+		cp->filter_dup = atomic_test_bit(bt_dev.flags,
+						 BT_DEV_SCAN_FILTER_DUP);
```
Code is just moved, will leave as is. Can be refactored afterwards.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -562,2355 +511,1756 @@ static int set_le_adv_enable_ext(struct bt_le_ext_adv *adv,
-		return 0;
+	if (enable == BT_HCI_LE_SCAN_ENABLE) {
+		cp->filter_dup = atomic_test_bit(bt_dev.flags,
+						 BT_DEV_SCAN_FILTER_DUP);
```
Code is just moved, will leave as is. Can be refactored afterwards.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -562,2355 +511,1756 @@ static int set_le_adv_enable_ext(struct bt_le_ext_adv *adv,
+int bt_le_scan_set_enable(uint8_t enable)
+{
+	if (IS_ENABLED(CONFIG_BT_EXT_ADV) &&
+	    BT_FEAT_LE_EXT_ADV(bt_dev.le.features)) {
```
Code is just moved, will leave as is. Can be refactored afterwards.

joerchan commented on subsys/bluetooth/host/hci_core.c:
Code is just moved, will leave as is. Can be refactored afterwards.

joerchan commented on subsys/bluetooth/host/hci_core.c:
Code is just moved, will leave as is. Can be refactored afterwards.

joerchan commented on subsys/bluetooth/host/hci_core.c:
We can take that separately. Preferably the move only copies files.

joerchan commented on subsys/bluetooth/host/id.c:
```diff
@@ -0,0 +1,1740 @@
+	    !atomic_test_bit(adv->flags, BT_ADV_USE_IDENTITY)) {
+		int err;
+
+		bt_le_adv_set_enable_ext(adv, false, NULL);
```
This code is just moved, but I don't think is much that could be done about the error. maybe just log it.
In any case we do something about it afterwards.

joerchan commented on subsys/bluetooth/host/id.c:
```diff
@@ -0,0 +1,1740 @@
+		    !atomic_test_bit(adv->flags, BT_ADV_USE_IDENTITY)) {
+			adv_enabled = true;
+			id = adv->id;
+			bt_le_adv_set_enable_legacy(adv, false);
```
Same as above.

joerchan commented on subsys/bluetooth/host/id.c:
```diff
@@ -0,0 +1,1740 @@
+	}
+
+done:
+	addr_res_enable(BT_HCI_ADDR_RES_ENABLE);
```
Missing check, but actually enabling and disabling of address resolution has no point. So this could be removed.
We can do that afterwards.

joerchan commented on subsys/bluetooth/host/id.c:
```diff
@@ -0,0 +1,1740 @@
+	    atomic_test_bit(bt_dev.flags, BT_DEV_ACTIVE_SCAN) &&
+	    !(IS_ENABLED(CONFIG_BT_EXT_ADV) &&
+	      atomic_test_bit(bt_dev.flags, BT_DEV_SCAN_LIMITED))) {
+		bt_le_scan_set_enable(BT_HCI_LE_SCAN_DISABLE);
```
Same as above. I think it is intentional in this case though.

joerchan commented:
@jhedberg @carlescufi How is it with the internal license header and the year? I just copied hci_core's license, but as @Thalley pointed out these year as very outdated. Do we really need the years? It would appear it is not that important to update the end year. Also what is the purpose of setting the copyright on the companies that edit the file, does it even make sense?
Couldn't it just be copyright Zephyrproject ?

jhedberg commented:
@joerchan ideally, the copyright year should be updated every time a copyright holder modifies the code for a new year. Note that (AFAIK, IANAL) copyright exists whether the declaration is there or not, i.e. the purpose of the declaration is to help out the reader to understand who the copyright owner is. The significance of the year would then be to determine when the copyright expires (i.e. the code becomes public domain). As for copyright ownership, that depends on your country's legislation and/or the employment contract. At least in Finland the copyright is implicitly owned by the employer for any work that even slightly relates to my employment. This is not something I have the authority to change.

Thalley reviewed (approved):
While this PR has exposed a lot of things that could be improved/changed, it doesn't seem like the changes (i.e. non-copy code blocks) have any noticeable issues, so looks good enough
